### DISCUSSION TITLE
Build error with MPAS-Ocean standalone

### DISCUSSION THREAD

**Discussion by daniloceano** [E3SM-Project/E3SM#5638] [E3SM model help]
This issue was originally open on the MPAS-Model repository and I was prompted to come here as, it seems, they no longer are involed with Mpas-Ocean development. 

I donwloaded the source code from https://mpas-dev.github.io/ and compiled the MPAS atmosphere model (everything ok on this part). Now I intend to work with the MPAS-Ocean but the Ocean core won't compile. Here's the fail message:

![image](https://user-images.githubusercontent.com/56005607/234091142-beb33bd0-3d09-482b-b5d1-71a258b6fb71.png)

And here is the full log error:

[log_build_ocean.txt](https://github.com/E3SM-Project/E3SM/files/11313768/log_build_ocean.txt)

Compilers used:
export SERIAL_FC=gfortran
export SERIAL_F77=gfortran
export SERIAL_CC=gcc
export SERIAL_CXX=g++
export MPI_FC=mpifort
export MPI_F77=mpifort
export MPI_CC=mpicc
export MPI_CXX=mpic++

Versions:
gcc version: 5.4.0
MPICH version 3.3.1

NetCDF
netcdf library version 4.7.0

Parallel NetCDF
PnetCDF Version: 1.11.2

PIO: 2 (couldn't find a way to check that but atmosphere_model was compiled using the flag USE_PIO2=true)

uname -a
Linux nemo 4.4.0-210-generic

https://github.com/MPAS-Dev/MPAS-Model/pull/242-Ubuntu SMP Fri Apr 16 09:57:56 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

make line used to build the executable:
make -j4 gfortran CORE=ocean PRECISION=double Debug=true USE_PIO2=true > log_build_ocean.txt

**Comment by vanroekel** [comment]
@daniloceano This code appears to be pretty dated and your issue appears to be a python 2 version of print that was fixed a long time ago.  Could you follow this rough set of steps instead

1. git clone git@github.com:E3SM-Project/E3SM.git
2. cd into E3SM
3. git submodule update --init --recursive
4. cd components/mpas-ocean
5. rebuild as you did for this issue.

This should get you passed the `makedep.py` error

Let me know if that works.

**Comment by daniloceano** [comment]
@vanroekel 

> 1. git clone [git@github.com](mailto:git@github.com):E3SM-Project/E3SM.git

This failed:

> ssh: connect to host github.com port 22: Connection timed out
fatal: Could not read from remote repository.
Please make sure you have the correct access rights
and the repository exists.

I assumed the path was just wrong and the cloned directly from https://github.com/E3SM-Project/E3SM. But it failed again when I prompted `git submodule update --init --recursive`:

> 
> Submodule 'externals/scorpio_classic' (git@github.com:E3SM-Project/scorpio.git) registered for path 'externals/scorpio_classic'
> Cloning into 'cime'...
> ssh: connect to host github.com port 22: Connection timed out
> fatal: Could not read from remote repository.
> 
> Please make sure you have the correct access rights
> and the repository exists.
> fatal: clone of 'git@github.com:ESMCI/cime.git' into submodule path 'cime' failed

I tried to build anyway:

`(ncl_stable) danilocs@nemo:/p1-nemo/danilocs/E3SM/components/mpas-ocean/src$ make -j4 gfortran CORE=ocean PRECISION=double Debug=true USE_PIO2=true > log_build_ocean.txt`

`make: *** No rule to make target 'gfortran'.  Stop.`

**Comment by vanroekel** [comment]
two thoughts
1. do you have a ssh key setup for github for the machine you are working on?  https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent

2. does the machine have some type of proxy that needs to be configured in the gitrc file?  e.g. https://gist.github.com/evantoli/f8c23a37eb3558ab8765****

E3SM should be a fully public repo so it should be clonable by anyone with a github account

**Comment by vanroekel** [comment]
sorry didn't fully read your comment.  The path I gave is correct but uses ssh instead of HTTPS.  I'm bguessing the latter two errors are similar and would suspect a proxy issue (my (2) above)

regarding the make issue you have, you are in the wrong directory, you are one level too deep.  you should be in mpas-ocean, not mpas-ocean/src

**Comment by daniloceano** [comment]
It is not possible for me to use ssh: the machine I am working won't allow using ssh to outside networks. I can clone the repo using HTTPS but can I do something similar to the "git submodule update --init --recursive" step but for the HTTPS? 
Sorry, I do not understand that much about git.

Anyway, I tried to compile the model without this step in case this is not necessary and it failed:

[log_build_ocean.txt](https://github.com/E3SM-Project/E3SM/files/11315156/log_build_ocean.txt)

**Comment by vanroekel** [comment]
Not sure what this error means, but am guessing it has to do with the incomplete checkout.  My sense is alterations to how the E3SM submodules are setup would require a lot of work that our project is not setup to do.  We only ensure E3SM will function on DOE computing platforms.  

I would suggest you try getting an account on NERSC computing -- https://docs.nersc.gov/accounts/ -- you can get what are called exploratory accounts (small computing allocations) relatively easily.  You will likely have an easier time getting MPAS-O / E3SM running there.  If you need to continue on the machine you are on you will have to work with your local computing support to sort this out.

**Comment by xylar** [comment]
@daniloceano, one more suggestion:
```
 git config --global url."https://github.com/".insteadOf "git@github.com:"
```
This should allow you to proceed with the recursive checkout.

**Comment by xylar** [comment]
I will warn you that MPAS-Ocean has very limited documentation and also very limited staff support for external users, so you may find that it isn't the right model for you unless you are willing to figure a lot of the steps out by looking through the code itself.

**Comment by daniloceano** [comment]
@vanroekel:
>I would suggest you try getting an account on NERSC computing -- https://docs.nersc.gov/accounts/ -- you can get what are called exploratory accounts (small computing allocations) relatively easily. You will likely have an easier time getting MPAS-O / E3SM running there. If you need to continue on the machine you are on you will have to work with your local computing support to sort this out.

This sounds a really good idea, thank you. If I may ask one more question... in order to create a NERSC acconunt I need to provide a project name for joinning, should it be the ES3M project itself?

@xylar:
`git config --global url."https://github.com/".insteadOf "git@github.com:"`

That worked, thank you. However, the model still won't run: 

```
(ncl_stable) danilocs@nemo:/p1-nemo/danilocs/E3SM/components/mpas-ocean$ make -j4 gfortran CORE=ocean PRECISION=double Debug=true USE_PIO2=true > log_build_ocean.txt
make[5]: warning: -jN forced in submake: disabling jobserver mode.
gfortran: error: pio2.f90: No such file or directory
gfortran: error: pio1.f90: No such file or directory
ar: `u' modifier ignored since `D' is the default (see `U')
ar: creating libesmf_time.a
make[7]: warning: -jN forced in submake: disabling jobserver mode.
make[7]: warning: -jN forced in submake: disabling jobserver mode.
make[6]: warning: -jN forced in submake: disabling jobserver mode.
make[6]: warning: -jN forced in submake: disabling jobserver mode.
mpas_io.F:280:41:

             pio_iotype = PIO_iotype_adios
                                         1
Error: Symbol ‘pio_iotype_adios’ at (1) has no IMPLICIT type
make[3]: *** [mpas_io.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [frame] Error 2
make[1]: *** [framework] Error 2
make: *** [gfortran] Error 2
```
The full log is here:

[log_build_ocean.txt](https://github.com/E3SM-Project/E3SM/files/11325732/log_build_ocean.txt)

> 
>I will warn you that MPAS-Ocean has very limited documentation and also very limited staff support for external users, so you may find that it isn't the right model for you unless you are willing to figure a lot of the steps out by looking through the code itself.

Thank you for the heads up. Isn't the model framework somewhat similar to the MPAS-A? This is one of the main reasons I opted for the MPAS-O...

**Comment by xylar** [comment]
@daniloceano, this is where we start to hit the limits fo the support we can provide.  There are several dependencies of MPAS-Ocean.  Once is PIO2 (we recommend the SCORPIO variant: https://github.com/E3SM-Project/scorpio).  You need to build this, point to it with the `PIO` environment variable, and make sure to either set the environment variable `USE_PIO2=true` or add that to your build command.

> Thank you for the heads up. Isn't the model framework somewhat similar to the MPAS-A? This is one of the main reasons I opted for the MPAS-O...

I could see why you might assume so but they're not all that similar, actually.  They used to use the same framework and they still use similar meshes, but the code bases have diverged and the build requirements and foci of the two projects were always different.  Being able to build and running MPAS-Atmosphere really doesn't prepare you for building and running MPAS-Ocean.  Also, they are developed by different institutions and with different levels of user support.
